### ISSUE TITLE
IIS Windows Authentication from a foreign (trusted) domain errors out

### ISSUE THREAD

**Issue by RobinBeismann** [Badgerati/Pode#534]
### Describe the Bug
IIS Windows Authentication does not work with users from a trusted domain.

### Steps To Reproduce
1. Setup Windows Authentication on IIS (doesn't matter if AppPoolIdentity or Domain Account)
2. Try to authenticate using a user from a trusted domain (transitive bi-directional trust)

### Expected Behavior
Users from trusted domains should be able to authenticate just like users from the local domain, in this setup it works for users from the local domain but not for users from the foreign domain.

### Platform
 - OS: Windows Server 2019
 - Browser Independent
 - Version: Latest PS 7.0, latest .NET Core Hosting Pack and Pode 1.7.0

### Additional Context
I am able to see the successful authentication from the foreign user on the Security Event Log, I can also logon to another application hosted on this IIS with Windows Authentication and the foreign user.
However I get an "401 - Failed to retrieve user from authentication token" from Pode.

I can quickly builtup a demo environment so we can debug this if you need.

This is the error I get when authenticating from a foreign user:
```
Date: 2020-04-16 00:23:40
Level: Error
ThreadId: 5
Server: %Servername%
Category: NotSpecified: (:) [Close-PodeDisposable], ExtendedTypeSystemException
Message: The following exception occurred while retrieving member "Dispose": "There is no such object on the server."
StackTrace: at Close-PodeDisposable, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Public\Utilities.ps1: line 90
at <ScriptBlock>, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Private\Authentication.ps1: line 475
at Invoke-PodeScriptBlock, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Public\Utilities.ps1: line 646
at <ScriptBlock>, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Private\Authentication.ps1: line 632
at Invoke-PodeScriptBlock, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Public\Utilities.ps1: line 638
at Invoke-PodeMiddleware, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Private\Middleware.ps1: line 46
at Invoke-PodeSocketHandler, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Private\PodeServer.ps1: line 233
at Invoke-PodeSocketProcessAccept, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Private\Sockets.ps1: line 202
at <ScriptBlock>, C:\Program Files\PowerShell\Modules\Pode\1.7.0\Private\Sockets.ps1: line 63
at <ScriptBlock>, <No file>: line 15
```

**Comment by RobinBeismann** [comment]
I also tested with -NoGroups and -NoLocalCheck at Add-PodeAuth, however that doesn't work either. I however realized that this speeds up scupPS pretty hard.

**Comment by RobinBeismann** [comment]
Ah I see, this must be related to the ADSI Query on Line 468 at Private\Authentication.ps1. I'll debug on our environment tomorrow.

**Comment by Badgerati** [comment]
I just tracked it down to the same ADSI query as well, it must be that the query LDAP and the User SID on the local domain only; so I wonder if there's a way to force the domain the query goes to 🤔 

If not, we could add a new `-NoADCheck` similar to `-NoLocalCheck` (if you don't care about the user's Email, Groups, Name, etc).

**Comment by RobinBeismann** [comment]
It would work this way, I'll try it later with Pode:

```
$sid = "sid"

function Get-ADSIKnownDomains(){
    $returnArray = @{}

    $searcher=[ADSIsearcher]'(objectclass=trustedDomain)'
    $searcher.searchroot.Path="LDAP://$($env:USERDNSDOMAIN)"
    
    #Search each of them        
    $searcher.FindAll() | ForEach-Object {
        $domainContext = New-Object -TypeName System.DirectoryServices.ActiveDirectory.DirectoryContext -ArgumentList ('Domain', $_.Properties.name)
        $domain = [DirectoryServices.ActiveDirectory.Domain]::GetDomain($domainContext)
        $returnArray[($_.Properties.flatname[0])] = $domain.GetDirectoryEntry()
    }
    #Add own domain
    $returnArray[$env:USERDOMAIN] = [ADSI]"LDAP://$($env:USERDNSDOMAIN)"
    #Return the result
    return $returnArray
}

$user = (Get-ADSIKnownDomains).GetEnumerator() | ForEach-Object {    
    $user = [adsisearcher]""
    $user.SearchRoot = [adsi]$($_.Value.Path)
    $user.Filter = "ObjectSid=$sid"

    $user.FindAll()
}
```

**Comment by Badgerati** [comment]
One of the things the WindowsIdentity gives us back is the domain for the user:

```powershell
$winIdentity = New-Object System.Security.Principal.WindowsIdentity($winAuthToken, 'Windows')

# get user and domain
$username = ($winIdentity.Name -split '\\')[-1]
$domain = ($winIdentity.Name -split '\\')[0]
```

Is there a chance we could use that domain, instead of looping through a lot adsi calls? (thinking of performance)

Possibly:

```powershell
$searcher = [adsisearcher]""
$searcher.SearchRoot = [adsi]"LDAP://$($domain)"
$searcher.Filter = "ObjectSid=$sid"
$user = $searcher.FindOne()
```

**Comment by RobinBeismann** [comment]
This does the trick. However the group functions don't work anymore.
I'll fix the group functions later today and raise a pull request.
```

                    $searcher = [adsisearcher]""
                    $searcher.SearchRoot = [adsi]"LDAP://$($domain)"
                    $searcher.Filter = "ObjectSid=$($winIdentity.User.Value.ToString())"

                    $ad = $searcher.FindOne()
                    
                    $user.DistinguishedName = @($ad.Properties.distinguishedname)[0]
                    $user.Name = @($ad.Properties.name)[0]
                    $user.Email = @($ad.Properties.mail)[0]
                    $user.Fqdn = (Get-PodeADServerFromDistinguishedName -DistinguishedName $user.Properties.DistinguishedName)
```

**Comment by RobinBeismann** [comment]
I've embedded and fixed this in #535.
Would it be possible to get this into a release like 1.7.1 so I can migrate my two productively used apps back to the PSGallery module instead of a local fork?

**Comment by Badgerati** [comment]
Yes, I will! I'll review the PR and test hopefully around lunch.

**Comment by Badgerati** [comment]
Released in v1.7.1
